crates/bevy_ecs/src/query/fetch.rs

```diff
@@ -440,6 +440,10 @@ pub unsafe trait WorldQuery {
     /// Creates and initializes a [`State`](WorldQuery::State) for this [`WorldQuery`] type.
     fn init_state(world: &mut World) -> Self::State;
 
+    /// Creates a [`State`](WorldQuery::State) for this [`WorldQuery`] type. This should return
+    /// the same value as [`init_state`](Self::init_state).
+    fn new_state(components: &Components) -> Option<Self::State>;
```

Comment:
This is a bit confusing to me, if "it should return the same value as 'init_state`"
- Why do we need both?
- How come one returns a `State` and one returns an `Option<State>`
- In what cases would the result be `None`?

Reply:
`init_state` also registers the component, which requires mutable access to the world. Which we don't have access to in a normal system. It'd return `None` when the component isn't registered. A better name for the method might be `get_state`